File: main/path-sum-iii/path-sum-iii.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def pathSum(self, root, total):
        """
        :type root: TreeNode
        :type total: int
        :rtype: int
        """
        if root is None:
            return 0

        path = []
        count = 0

        def doCount(partial):
            nonlocal count

            if partial == 0:
                count += 1
            for val in path:
                partial += val
                if partial == 0:
                    count += 1

        def dfs(node):
            nonlocal total

            total -= node.val
            doCount(total)
            path.append(node.val)

            if node.left is not None:
                dfs(node.left)
            if node.right is not None:
                dfs(node.right)

            total += path.pop()

        dfs(root)
        return count

File: main/path-sum-iii/path-sum-iii.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public int PathSum(TreeNode root, int sum) {
        if (root == null) return 0;
        var path = new List<int>();
        var count = 0;

        void DoCount(int partial) {
            if (partial == 0) ++count;

            foreach (var val in path) {
                partial += val;
                if (partial == 0) ++count;
            }
        }

        void Dfs(TreeNode node) {
            sum -= node.val;
            DoCount(sum);
            path.Add(node.val);

            if (node.left != null) Dfs(node.left);
            if (node.right != null) Dfs(node.right);

            path.RemoveAt(path.Count - 1);
            sum += node.val;
        }

        Dfs(root);
        return count;
    }
}
